### PR TITLE
Feature/add array replace nth function

### PR DESCRIPTION
**Pull request by dothebart**
### Scope & Purpose

This PR adds a new AQL function which produces an array that replaces one value with a specified new value. This could previously only achieved by combining several SLICE() operations.

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

#### Related Information
- [x] There is a *JIRA Ticket number* (In case a customer was affected / involved): DEVSUP-540 (JPMC)

### Testing & Verification
This PR adds tests that were used to verify all changes:
- [x] Added new **integration tests** in shell_server_aql

### Documentation
- [x] Added a *Changelog Entry* (referencing the corresponding public or internal issue number)
- [ ] Added a new section in the *Manual*

### CODE REVIEW

**Comment by dothebart** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8206/

**Comment by dothebart** [comment]
test green.

File: arangod/Aql/Functions.cpp [hunk 1]
```diff
@@ -6414,6 +6414,67 @@ AqlValue Functions::RemoveNth(ExpressionContext* expressionContext,
   return AqlValue(builder.get());
 }
 
+/// @brief function ReplaceNth
+AqlValue Functions::ReplaceNth(ExpressionContext* expressionContext, transaction::Methods* trx,
+                            VPackFunctionParameters const& parameters) {
+  // cppcheck-suppress variableScope
+  static char const* AFN = "RELPLACE_NTH";
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]:
```suggestion
  static char const* AFN = "REPLACE_NTH";
```

File: arangod/Aql/Functions.cpp [hunk 2]
```diff
@@ -6414,6 +6414,67 @@ AqlValue Functions::RemoveNth(ExpressionContext* expressionContext,
   return AqlValue(builder.get());
 }
 
+/// @brief function ReplaceNth
+AqlValue Functions::ReplaceNth(ExpressionContext* expressionContext, transaction::Methods* trx,
+                            VPackFunctionParameters const& parameters) {
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: indentation?

File: arangod/Aql/Functions.cpp [hunk 3]
```diff
@@ -6414,6 +6414,67 @@ AqlValue Functions::RemoveNth(ExpressionContext* expressionContext,
...
+  AqlValue const& paddValue = extractFunctionParameterValue(parameters, 3);
+
+  bool havePadValue = parameters.size() == 4;
+
+  uint64_t replaceOffset;
+  if (offset.isNull(true) || offset.toInt64() < 0) {
+    THROW_ARANGO_EXCEPTION_PARAMS(TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH, AFN);
+  } else {
+    replaceOffset = static_cast<uint64_t>(offset.toInt64());
+  }
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: block can be simplified to
```

  if (offset.isNull(true) || offset.toInt64() < 0) {
    THROW_ARANGO_EXCEPTION_PARAMS(TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH, AFN);
  }
  uint64_t replaceOffset = static_cast<uint64_t>(offset.toInt64());
```

File: arangod/Aql/Functions.cpp [hunk 4]
```diff
@@ -6414,6 +6414,67 @@ AqlValue Functions::RemoveNth(ExpressionContext* expressionContext,
...
+  
+  if (!baseArray.isArray()) {
+    registerInvalidArgumentWarning(expressionContext, AFN);
+    return AqlValue(AqlValueHintNull());
+  }
+  
+  if ((baseArray.length() < replaceOffset ) &&!havePadValue) {
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]:
```suggestion
  if (baseArray.length() < replaceOffset && !havePadValue) {
```

File: arangod/Aql/Functions.cpp [hunk 5]
```diff
@@ -6414,6 +6414,67 @@ AqlValue Functions::RemoveNth(ExpressionContext* expressionContext,
...
+  VPackSlice replaceValue = materializer.slice(newValue, false);
+  
+  transaction::BuilderLeaser builder(trx);
+  builder->openArray();
+  
+
+  uint64_t pos = 0;
+  VPackArrayIterator it(arraySlice);
+  while (it.valid()) {
+    if (pos != replaceOffset) {
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]:
```suggestion
    if (it.index() != replaceOffset) {
```

File: arangod/Aql/Functions.cpp [hunk 6]
```diff
@@ -6414,6 +6414,67 @@ AqlValue Functions::RemoveNth(ExpressionContext* expressionContext,
...
+
+  AqlValueMaterializer materializer(trx);
+  VPackSlice arraySlice = materializer.slice(baseArray, false);
+  VPackSlice replaceValue = materializer.slice(newValue, false);
+  
+  transaction::BuilderLeaser builder(trx);
+  builder->openArray();
+  
+
+  uint64_t pos = 0;
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: not needed

File: tests/js/server/aql/aql-functions.js [hunk 7]
```diff
@@ -880,6 +880,59 @@ function ahuacatlFunctionsTestSuite () {
...
+        [],
+        [7, 6],
+        { },
+        { foo: 'bar'}
+      ];
+
+      for (let replaceIndex = 0;  replaceIndex <= testArray.length + 2; replaceIndex ++) {
+        for (let replaceValue = 0;  replaceValue < testArray.length; replaceValue ++) {
+          const msg = `Index: ${replaceIndex} ReplaceValue: ${replaceValue}`;
+          actual = getQueryResults("RETURN NOOPT(REPLACE_NTH(@testArray, @which, @replaceValue, @defaultValue))",
```
Comment (jsteemann) [inline tests/js/server/aql/aql-functions.js]:
```suggestion
          let actual = getQueryResults("RETURN NOOPT(REPLACE_NTH(@testArray, @which, @replaceValue, @defaultValue))",
```

File: tests/js/server/aql/aql-functions.js [hunk 8]
```diff
@@ -880,6 +880,59 @@ function ahuacatlFunctionsTestSuite () {
       assertEqual([ [ ] ], getQueryResults("RETURN NOOPT(SLICE([ ], 1, [ ]))")); 
     },
 
+////////////////////////////////////////////////////////////////////////////////
+/// @brief test replace_nth function
+////////////////////////////////////////////////////////////////////////////////
+
+    testReplaceNthCxx : function () {
+      var actual;
```
Comment (jsteemann) [inline tests/js/server/aql/aql-functions.js]:
```suggestion
```

**Comment by jsteemann** [comment]
Please also open a PR for the docs repository with the changes (for the AQL functions section and the new features page).

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8287/

**Pull request merged**